akka-javasdk-testkit/src/main/java/akka/javasdk/testkit/TelemetryReader.java

```diff
@@ -0,0 +1,183 @@
+/*
+ * Copyright (C) 2021-2025 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.javasdk.testkit;
+
+import static java.util.Optional.ofNullable;
+
+import akka.annotation.ApiMayChange;
+import akka.javasdk.annotations.Component;
+import akka.javasdk.annotations.ComponentId;
+import akka.javasdk.workflow.Workflow;
+import io.opentelemetry.api.common.AttributeKey;
+import io.opentelemetry.sdk.trace.data.SpanData;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Optional;
+import kalix.runtime.telemetry.tracing.TracingSetup;
+
+/**
+ * A test utility for reading and inspecting telemetry data captured during test execution.
+ *
+ * <p>TelemetryReader provides access to OpenTelemetry spans collected by the in-memory span
+ * exporter during integration tests. This allows you to verify workflow execution, agent
+ * interactions, and other instrumented operations.
+ */
+@ApiMayChange
+public class TelemetryReader {
+
+  private final TracingSetup.AkkaInMemorySpanExporter inMemorySpanExporter;
+
+  public TelemetryReader(TracingSetup.AkkaInMemorySpanExporter inMemorySpanExporter) {
+    this.inMemorySpanExporter = inMemorySpanExporter;
+  }
+
+  /**
+   * Retrieves the sequence of workflow steps executed for a specific workflow instance.
+   *
+   * <p>Returns a list of step names in the order they were executed, based on telemetry data
+   * collected during the workflow execution. This is useful for verifying that workflows execute
+   * the expected steps in the correct order.
+   *
+   * @param workflow The workflow class to query
+   * @param workflowId The unique identifier of the workflow instance
+   * @return A list of step names in execution order, or an empty list if no steps were found
+   */
+  @ApiMayChange
+  public List<String> getWorkflowSteps(Class<? extends Workflow<?>> workflow, String workflowId) {
+    String componentId = getComponentId(workflow);
+    List<SpanData> spanDatas =
+        spansFor(componentId, AttributeKey.stringKey("akka.workflow.id"), workflowId);
+
+    return workflowStepsFrom(spanDatas);
+  }
+
+  /**
+   * Retrieves the list of agents that were invoked during an operation.
+   *
+   * <p>Returns a list of agent IDs in the order they were invoked, based on telemetry data
+   * collected during test execution. This is useful for verifying that the expected agents were
+   * called during a test scenario.
+   *
+   * @param debugId The debug identifier used to trace the operation
+   * @return A list of agent IDs in invocation order, or an empty list if no agents were found
+   */
+  @ApiMayChange
+  public List<String> getAgents(String debugId) {
```

Comment:
Should there be a look up by session id? Or accessing agent spans that are part of a trace identified in another way (such as by workflow)?

Reply:
I would say both options are interesting. When testing two separate requests I think debugId is quite useful. Otherwise, I would get an accumulated invocations because it is the same workflow id. For other cases/tests the second option might be more useful. I could extend the API with the workflowId.

Reply:
OK, I have done some experiments with multi-agent sample and it's a bit confusing for myself to write a good assertion using this method, basically the only one that makes sense to me is at the end of the test, but it's a bit redundant because of previous check and I don't see a much value from it. I think we could always add it later. I was also thinking about extending the API and return sth else than just string, for instance steps with payload, etc.

Reply:
Yeah, having more detail than just the names would be useful. And could also separate out into finding the trace data that you're looking for (could be matching on different types of attributes, including debug id), and then matching on the data in that trace.